# Activity 3.1 - Recurrent Neural Networks

#### Objective(s):

This activity aims to introduce how to build a recurrent neural network

#### Intended Learning Outcomes (ILOs):
* Demonstrate how to build and train neural recurrent neural network
* Evaluate the score and accuracy of the recurrent neural network

#### Resources:
* Jupyter Notebook
* IMDB

#### Procedures
Load the necessary libraries

In [1]:
from __future__ import print_function
import keras
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import SimpleRNN
from keras.datasets import imdb
from keras import initializers

* Set the max_features to 20000
* Set the maximum length of a sequence
* Use the batch size of 32

In [2]:
max_features = 20000
maxlen = 30
batch_size = 32

Load the data

In [3]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

17464789/17464789 [==============================] - 1s 0us/step
25000 train sequences
25000 test sequences


 Truncates the sequences so that they are of the maximum length

In [4]:
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (25000, 30)
x_test shape: (25000, 30)


In [7]:
# Check the example of sequence

In [8]:
x_train[123,:]

array([  219,   141,    35,   221,   956,    54,    13,    16,    11,
        2714,    61,   322,   423,    12,    38,    76,    59,  1803,
          72,     8, 10508,    23,     5,   967,    12,    38,    85,
          62,   358,    99], dtype=int32)

Build a recurrent neural network

In [9]:
rnn_hidden_dim = 5
word_embedding_dim = 50
model_rnn = Sequential()
model_rnn.add(Embedding(max_features, word_embedding_dim))  #This layer takes each integer in the sequence and embeds it in a 50-dimensional vector
model_rnn.add(SimpleRNN(rnn_hidden_dim,
                    kernel_initializer=initializers.RandomNormal(stddev=0.001),
                    recurrent_initializer=initializers.Identity(gain=1.0),
                    activation='relu',
                    input_shape=x_train.shape[1:]))

model_rnn.add(Dense(1, activation='sigmoid'))

In [10]:
rmsprop = keras.optimizers.RMSprop(lr = .0001)

model_rnn.compile(loss='binary_crossentropy',
              optimizer=rmsprop,
              metrics=['accuracy'])

In [11]:
model_rnn.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=10,
          validation_data=(x_test, y_test))

Epoch 1/10
782/782 [==============================] - 40s 47ms/step - loss: 0.6030 - accuracy: 0.6872 - val_loss: 0.5794 - val_accuracy: 0.7176
Epoch 2/10
782/782 [==============================] - 29s 37ms/step - loss: 0.5014 - accuracy: 0.7799 - val_loss: 0.5024 - val_accuracy: 0.7692
Epoch 3/10
782/782 [==============================] - 30s 38ms/step - loss: 0.4501 - accuracy: 0.8069 - val_loss: 0.5381 - val_accuracy: 0.7600
Epoch 4/10
782/782 [==============================] - 31s 40ms/step - loss: 0.4143 - accuracy: 0.8215 - val_loss: 0.4716 - val_accuracy: 0.7782
Epoch 5/10
782/782 [==============================] - 29s 37ms/step - loss: 0.3862 - accuracy: 0.8379 - val_loss: 0.4908 - val_accuracy: 0.7766
Epoch 6/10
782/782 [==============================] - 28s 36ms/step - loss: 0.3599 - accuracy: 0.8500 - val_loss: 0.4722 - val_accuracy: 0.7799
Epoch 7/10
782/782 [==============================] - 29s 37ms/step - loss: 0.3385 - accuracy: 0.8626 - val_loss: 0.4866 - val_accuracy:

Evaluate the model using the test set.

In [12]:
score, acc = model_rnn.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

782/782 [==============================] - 4s 5ms/step - loss: 0.5145 - accuracy: 0.7686
Test score: 0.5144986510276794
Test accuracy: 0.7686399817466736


Interpret the result

#type your answer here

The RNN model achieves a test accuracy of about 76.86%. This means it's performing reasonably well, but there's still room for improvement. The loss of 0.5145 indicates that there's a moderate amount of error in the predictions.

#### Supplementary Activity

- Prepare the data to use sequences of length 80 rather than length 30.  Did it improve the performance?
- Try different values of the "max_features".  Can you improve the performance?
- Try smaller and larger sizes of the RNN hidden dimension.  How does it affect the model performance?  How does it affect the run time?

#type your answer here

# #1

In [30]:
# type your code here
#different values of max_features and maxlen
max_features = 10000
maxlen = 80
batch_size = 32

In [31]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

25000 train sequences
25000 test sequences


In [32]:
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (25000, 80)
x_test shape: (25000, 80)


In [33]:
# Trying smaller size of RNN hidden dimension
rnn_hidden_dim = 2
word_embedding_dim = 50
supp_model1 = Sequential()
supp_model1.add(Embedding(max_features, word_embedding_dim))  #This layer takes each integer in the sequence and embeds it in a 50-dimensional vector
supp_model1.add(SimpleRNN(rnn_hidden_dim,
                    kernel_initializer=initializers.RandomNormal(stddev=0.001),
                    recurrent_initializer=initializers.Identity(gain=1.0),
                    activation='relu',
                    input_shape=x_train.shape[1:]))

supp_model1.add(Dense(1, activation='sigmoid'))

In [34]:
rmsprop = keras.optimizers.RMSprop(lr = .0001)

supp_model1.compile(loss='binary_crossentropy',
              optimizer=rmsprop,
              metrics=['accuracy'])

In [35]:
supp_model1.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=10,
          validation_data=(x_test, y_test))

Epoch 1/10
782/782 [==============================] - 84s 106ms/step - loss: 0.5826 - accuracy: 0.7122 - val_loss: 0.5095 - val_accuracy: 0.8073
Epoch 2/10
782/782 [==============================] - 71s 91ms/step - loss: 0.4421 - accuracy: 0.8337 - val_loss: 0.5156 - val_accuracy: 0.7469
Epoch 3/10
782/782 [==============================] - 63s 81ms/step - loss: 0.3741 - accuracy: 0.8580 - val_loss: 0.4905 - val_accuracy: 0.8034
Epoch 4/10
782/782 [==============================] - 65s 83ms/step - loss: 0.3294 - accuracy: 0.8750 - val_loss: 0.5142 - val_accuracy: 0.7411
Epoch 5/10
782/782 [==============================] - 67s 86ms/step - loss: 0.3003 - accuracy: 0.8872 - val_loss: 0.7014 - val_accuracy: 0.7697
Epoch 6/10
782/782 [==============================] - 70s 89ms/step - loss: 0.2773 - accuracy: 0.8951 - val_loss: 0.3992 - val_accuracy: 0.8279
Epoch 7/10
782/782 [==============================] - 68s 87ms/step - loss: 0.2576 - accuracy: 0.9057 - val_loss: 0.4260 - val_accuracy

In [36]:
score, acc = supp_model1.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

782/782 [==============================] - 7s 9ms/step - loss: 0.4665 - accuracy: 0.8002
Test score: 0.4664912223815918
Test accuracy: 0.8001999855041504


In [37]:
# Trying larger size of RNN hidden dimension
rnn_hidden_dim = 10
word_embedding_dim = 50
supp_model2 = Sequential()
supp_model2.add(Embedding(max_features, word_embedding_dim))  #This layer takes each integer in the sequence and embeds it in a 50-dimensional vector
supp_model2.add(SimpleRNN(rnn_hidden_dim,
                    kernel_initializer=initializers.RandomNormal(stddev=0.001),
                    recurrent_initializer=initializers.Identity(gain=1.0),
                    activation='relu',
                    input_shape=x_train.shape[1:]))

supp_model2.add(Dense(1, activation='sigmoid'))

In [38]:
rmsprop = keras.optimizers.RMSprop(lr = .0001)

supp_model2.compile(loss='binary_crossentropy',
              optimizer=rmsprop,
              metrics=['accuracy'])

supp_model2.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=10,
          validation_data=(x_test, y_test))

Epoch 1/10
782/782 [==============================] - 74s 94ms/step - loss: 0.4713 - accuracy: 0.7577 - val_loss: 0.3926 - val_accuracy: 0.8211
Epoch 2/10
782/782 [==============================] - 67s 86ms/step - loss: 0.3260 - accuracy: 0.8584 - val_loss: 0.3645 - val_accuracy: 0.8405
Epoch 3/10
782/782 [==============================] - 70s 90ms/step - loss: 0.2899 - accuracy: 0.8798 - val_loss: 0.3582 - val_accuracy: 0.8428
Epoch 4/10
782/782 [==============================] - 63s 80ms/step - loss: 0.2682 - accuracy: 0.8896 - val_loss: 0.3720 - val_accuracy: 0.8327
Epoch 5/10
782/782 [==============================] - 63s 80ms/step - loss: 0.2507 - accuracy: 0.8989 - val_loss: 0.3766 - val_accuracy: 0.8409
Epoch 6/10
782/782 [==============================] - 65s 83ms/step - loss: 0.2341 - accuracy: 0.9069 - val_loss: 0.3952 - val_accuracy: 0.8332
Epoch 7/10
782/782 [==============================] - 67s 85ms/step - loss: 0.2221 - accuracy: 0.9114 - val_loss: 0.3867 - val_accuracy:

In [39]:
score, acc = supp_model2.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

782/782 [==============================] - 8s 10ms/step - loss: 0.4419 - accuracy: 0.8280
Test score: 0.44188082218170166
Test accuracy: 0.8279600143432617


#### Conclusion

#type your answer here



Recurrent Neural Networks (RNNs) are powerful tools for analyzing sequential data like text. After doing the supplementary activity, increasing sequence length captures longer-term dependencies but can increase training time. Decreasing max_features simplifies the vocabulary, potentially aiding generalization. Increasing the hidden dimension allows the model to learn more complex patterns but may lead to overfitting. However, in the case of this activity, the test accuracy and score improves.